enable bf16 mkldnn path for gemm

zhuhaozhe opened the pull request:
# Goal: Integrate mkldnn bf16 Gemm to pytorch

## BF16 Suport for mm, addmm, bmm, addbmm, baddbmm, mv, addmv, dot (with mkldnn matmul primitive):
https://oneapi-src.github.io/oneDNN/group__dnnl__api__matmul.html
For gemm related ops, we keep all inputs under plain format. So we will not introduce opaque tensor for these ops to save mem copy here.

![mkldnn bf16 gemm integration](https://user-images.githubusercontent.com/54701539/126263077-4b5134e1-52a7-4fad-94fb-19e13a0377f6.png)

The minimized integration is only dispatch to mkldnn in addmm, but for gemm with 3-D input (with additional dim for"batch") this will call mkldnn gemm for "batch" times. Since mkldnn matmul support input with multiple dims, we directly dispatch to mkldnn gemm in {bmm, addbmm, baddbmm} to reduce the time to create mkldnn memory desc, primitive, etc.

For the different definition for "bias" between mkldnn(which must be shape of (1, N)) and pytorch (which can be same shape with gemm result (M, N)), we use a fused sum to handle it.

## User Case:
User case is exactly same with before because no opaque tensor's is introduced. Since the pytorch has already support bf16 data type with CPU tensor before, we can leverage the existed bf16 gemm UT.

## Gemm performance gain on CPX 28Cores/Socket:
Note: data is collected using PyTorch operator benchmarks: https://github.com/pytorch/pytorch/tree/master/benchmarks/operator_benchmark (with adding bfloat16 dtype)

### use 1 thread on 1 core
### torch.addmm (M, N) * (N, K) + (M, K) 
| impl |16x16x16|32x32x32| 64x64x64 | 128x128x128| 256x256x256| 512x512x512|1024x1024x1024|
|:---:|:---:| :---: | :---: | :---: | :---: | :---: | :---: |
| aten-fp32| 4.115us|4.583us|8.230us|26.972us|211.857us|1.458ms|11.258ms|
| aten-bf16 | 15.812us| 105.087us|801.787us|3.767ms|20.274ms|122.440ms|836.453ms|
| mkldnn-bf16 |20.561us |22.510us|24.551us|37.709us|143.571us|0.835ms|5.76ms|  

We can see mkldnn-bf16 are better than aten bf16, but for smaller shapes, mkldnn bf16 are not better than aten fp32. This is because onednn overhead, this overhead more like a "constant" overhead and while problems get larger, we can ignore it. Also we are continue optimize the kernel efficiency and decrease the overhead as well. 

More shapes
| impl |1x2048x2048|2048x1x2048| 2048x2048x1 | 
|:---:|:---:| :---: | :---: |
| aten-fp32| 0.640ms|3.794ms|0.641ms|
| aten-bf16 | 2.924ms| 3.868ms|23.413ms|
| mkldnn-bf16 |0.335ms |4.490ms|0.368ms|

### use 1 socket (28 thread, 28 core)
| impl | 256x256x256| 512x512x512|1024x1024x1024| 2048x2048x2048|4096x4096x4096|
|:---:| :---: | :---: | :---: | :---: | :---: |
| aten-fp32| 35.943us |140.315us|643.510us|5.827ms|41.761ms|
| mkldnn-bf16 |53.432us|114.716us|421.858us|2.863ms|23.029ms|

More shapes
| impl |128x2048x2048|2048x128x2048| 2048x2048x128 | 
|:---:|:---:| :---: | :---: |
| aten-fp32| 0.561ms|0.458ms|0.406ms|
| mkldnn-bf16 |0.369ms |0.331ms|0.239ms|

We dose not show aten-bf16 for this case since aten-bf16 always compute as single thread and the performance is extreme poor. The trend for this case is similar for 1 thread on 1 core.


cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @VitalyFedyunin

zhuhaozhe committed:
enable bf16 mkldnn path for gemm

zhuhaozhe commented:
Hi, @mingfeima @jgong5 Please help to review

mingfeima commented on aten/src/ATen/native/mkldnn/MKLDNNCommon.cpp:
```diff
@@ -86,12 +86,22 @@ ideep::tensor itensor_view_from_dense(const Tensor& tensor) {
+            {tensor.strides().cbegin(), tensor.strides().cend()}},
+            tensor.template data_ptr<float>()};
+  }
+  else{
```
indent

zhuhaozhe replied:
Thanks, changed

mingfeima commented on aten/src/ATen/native/mkldnn/Matmul.cpp:
```diff
@@ -0,0 +1,75 @@
+  TORCH_CHECK(mat2.dim() == 2 || mat2.dim() == 3, "mkldnn_matmul:  expect mat2 to be 2-D or 3-D tensor");
+  TORCH_CHECK(mat1.scalar_type() == at::kBFloat16, "mkldnn_matmul:  only enabled for bf16 path");
+  TORCH_CHECK(mat2.scalar_type() == at::kBFloat16, "mkldnn_matmul:  only enabled for bf16 path");
+  TORCH_CHECK(result.scalar_type() == at::kBFloat16, "mkldnn_matmul:  only enabled for bf16 path");
```
shouldn't we check
```
TORCH_CHECK((mat1.dim() ==  2 && mat2.dim() == 2) || (mat1.dim() == 3 && mat2.dim() == 3))
```

zhuhaozhe replied:
Changed.

mingfeima commented on aten/src/ATen/native/mkldnn/Matmul.cpp:
```diff
@@ -0,0 +1,75 @@
+  if (alpha == 0)
+    return;
+  Tensor mat1_ = mat1.unsafeGetTensorImpl()->is_wrapped_number() ? mat1.contiguous() : mat1;
+  Tensor mat2_ = mat1.unsafeGetTensorImpl()->is_wrapped_number() ? mat2.contiguous() : mat2;
```
first i don't quite understand why we need to call `unsafeGetTensorImpl()->is_wrapped_number()` here.
second `mat.contiguous()` will create additional copy for the case when the tensor is 'transposed' contiguous, any problem if we directly create ideep::tensor based on size and stride for this case?

zhuhaozhe replied:
It is because mkldnn dose not optimize these formats and will run into "ref" path.

mingfeima commented on aten/src/ATen/native/mkldnn/Matmul.cpp:
```diff
@@ -0,0 +1,75 @@
+  TORCH_CHECK(result.scalar_type() == at::kBFloat16, "mkldnn_matmul:  only enabled for bf16 path");
+  TORCH_CHECK(mkldnn_bf16_device_check(),
+    "mkldnn_matmul: mkldnn_matmul bf16 path needs the cpu support avx512bw, avx512vl and avx512dq");
+  // In aten, expect beta * result = 0 if beta = 0
```
you'd better give an explaination of API difference of mkldnn and pytorch matmul, aka. why we need post op here.

mingfeima commented on aten/src/ATen/native/mkldnn/Utils.h:
```diff
@@ -21,4 +22,29 @@ inline bool mkldnn_bf16_device_check() {
+  c10::MaybeOwned<Tensor> result_maybe_owned = at::borrow_from_optional_tensor(result_opt);
+  const Tensor& result = *result_maybe_owned;
+
+  const int64_t mkldnn_gemm_min_size = 32 * 32 *32;
```
the given benchmark shows even for 32^3, mkldnn bf16 is faster then ATen native impl, just it is slower than aten fp32 (which is mkl), why we need such a threshold here?

also check libxsmm bf16 perf to double confirm whether mkldnn counterpart need to be further improved.

zhuhaozhe replied:
Changed it to 16^3

mingfeima commented:
@zhuhaozhe, follow ups:
* provide benchmark data on latest machine with AMX, if possible.
* benchmark non-square shapes, e.g. thin, flat. Suppose gemm shape of Linear when bs=1 for commonly used topos.
* cross compare mkldnn with libxsmm (if you have the bandwidth, we can always push mkldnn to limit later on).

zhuhaozhe commented:
> @zhuhaozhe, follow ups:
> 
> * provide benchmark data on latest machine with AMX, if possible.
> * benchmark non-square shapes, e.g. thin, flat. Suppose gemm shape of Linear when bs=1 for commonly used topos.
> * cross compare mkldnn with libxsmm (if you have the bandwidth, we can always push mkldnn to limit later on).

Thanks, Mingfei.
Have added test results on SPR and more kinds of shapes.
These numbers are collected based on the current mkldnn version in ideep,  and it will be better after upgrade onednn version.

ezyang commented:
@mdschatz are you planning on reviewing this?

VitalyFedyunin commented on aten/src/ATen/native/mkldnn/Utils.h:
```diff
@@ -21,4 +22,29 @@ inline bool mkldnn_bf16_device_check() {
+  int64_t m = mat1.dim() == 2? mat1.size(0) : mat1.size(1);
+  int64_t n = mat1.dim() == 2? mat1.size(1) : mat1.size(2);
+  int64_t k = mat2.dim() == 2? mat2.size(1) : mat2.size(2);
+  return (
```
Please move fast checks (for example type) to the beginning of the function.

VitalyFedyunin commented:
1) Please update ideep as separate PR and mention why you need to move it
2) Please rebase

zhuhaozhe commented:
>     1. Please update ideep as separate PR and mention why you need to move it
> 
>     2. Please rebase

Hi, Vitaly. I have removed the dependency for ideep and rebased. Thanks.
Basically, I need update ideep before because in some cases. Aten will give me an result buffer which is not expected by onednn and I choose to handle it in ideep. Now I  handle it in pytorch.

zhuhaozhe committed:
remove dependency for ideep

zhuhaozhe committed:
move fast checks to the beginning

Jianhui-Li commented:
@VitalyFedyunin Could you please review this one also and see whether we can get in before the branch cut?  It doesn't depend on ideep/oneDNN upgrade.

VitalyFedyunin commented:
Can you please rebase to clear windows builds, failures loooks unrelated, but to be 100% sure.

zhuhaozhe commented:
> Can you please rebase to clear windows builds, failures loooks unrelated, but to be 100% sure.

Hi, Vitaly. I have rebased it,

VitalyFedyunin reviewed (approved):
It makes sense to benchmark it further for smaller shapes and find the sizes when mkldnn gemm becomes faster

VitalyFedyunin commented on tools/build_variables.bzl:
```diff
@@ -855,6 +855,8 @@ aten_cpu_source_non_codegen_list = [
     "aten/src/ATen/native/mkldnn/UnaryOps.cpp",
     "aten/src/ATen/native/mkldnn/Utils.cpp",
+    "aten/src/ATen/native/mkldnn/Matmul.cpp",
+    "aten/src/ATen/native/mkldnn/Matmul.h",
```
Please remove .h file from this list.

VitalyFedyunin reviewed (commented):
Please rebase the PR (plus do tiny change that I had locally). I'm still trying to merge it.

zhuhaozhe committed:
remote .h file from list

zhuhaozhe commented:
> Please rebase the PR (plus do tiny change that I had locally). I'm still trying to merge it.

Thanks, have removed the ".h" and rebased.

ngimel commented:
Reverted, it likely leads to rocm failures https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm4.2-py3.6-test1/17850//consoleFull

VitalyFedyunin commented:
Agree it seems to be related @zhuhaozhe please re-create PR and check why `itensor_view_from_dense` is used in CUDA (ROCM) test `test_addbmm_cuda_bfloat16`

mruberry commented:
This might also have caused win-vs2019-cuda10.1-py3 to fail with

```
======================================================================
ERROR [0.815s]: test_baddbmm_cuda_float16 (__main__.TestLinalgCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\actions-runner\_work\pytorch\pytorch\pytorch-1211232028\build\win_tmp\build\torch\testing\_internal\common_device_type.py", line 377, in instantiated_test
    raise rte
  File "C:\actions-runner\_work\pytorch\pytorch\pytorch-1211232028\build\win_tmp\build\torch\testing\_internal\common_device_type.py", line 372, in instantiated_test
    result = test(self, **param_kwargs)
  File "C:\actions-runner\_work\pytorch\pytorch\pytorch-1211232028\build\win_tmp\build\torch\testing\_internal\common_device_type.py", line 948, in only_fn
    return fn(self, *args, **kwargs)
  File "C:\actions-runner\_work\pytorch\pytorch\pytorch-1211232028\build\win_tmp\build\torch\testing\_internal\common_cuda.py", line 143, in wrapped
    f(**kwargs)
  File "test_linalg.py", line 6488, in test_baddbmm
    self._test_addbmm_baddbmm("baddbmm", b1, b2, ref, out_tensor)
  File "test_linalg.py", line 6311, in _test_addbmm_baddbmm
    self.assertEqual(out_tensor, ref)
  File "C:\actions-runner\_work\pytorch\pytorch\pytorch-1211232028\build\win_tmp\build\torch\testing\_internal\common_utils.py", line 1557, in assertEqual
    result, debug_msg_generic = self._compareTensors(x, y, rtol=rtol, atol=atol,
  File "C:\actions-runner\_work\pytorch\pytorch\pytorch-1211232028\build\win_tmp\build\torch\testing\_internal\common_utils.py", line 1425, in _compareTensors
    return _compare_tensors_internal(a, b, rtol=rtol, atol=atol, equal_nan=equal_nan)
  File "C:\actions-runner\_work\pytorch\pytorch\pytorch-1211232028\build\win_tmp\build\torch\testing\_core.py", line 96, in _compare_tensors_internal
    if torch.allclose(a, b, rtol=rtol, atol=atol, equal_nan=equal_nan):
RuntimeError: CUDA error: misaligned address
CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
```

so there may be a separate CUDA issue with the updated test sizes, too. cc @ngimel @xwang233

ngimel commented:
cc @zasdfgbnm, test sizes were changed in this PR, this could have legitimately lead to misaligned address if there's a cublas bug?

zhuhaozhe commented:
Hi, @VitalyFedyunin. I re-created PR https://github.com/pytorch/pytorch/pull/64619. The CUDA BF16 fail seems because I have not checked the Tensor device. But I do not know the reason for  CUDA FP16 fails now.
